binding.go

```diff
@@ -202,8 +205,23 @@ func Json(jsonStruct interface{}, ifacePtr ...interface{}) macaron.Handler {
 		ensureNotPointer(jsonStruct)
 		jsonStruct := reflect.New(reflect.TypeOf(jsonStruct))
 		if ctx.Req.Request.Body != nil {
-			defer ctx.Req.Request.Body.Close()
-			err := json.NewDecoder(ctx.Req.Request.Body).Decode(jsonStruct.Interface())
+			var err error
+			if ctx.Req.Method == "POST" || ctx.Req.Method == "PUT" || ctx.Req.Method == "PATCH" {
+				defer ctx.Req.Request.Body.Close()
+				v := jsonStruct.Interface()
+				if pb, ok := v.(proto.Message); ok {
+					err = jsonpb.Unmarshal(ctx.Req.Request.Body, pb)
+				} else {
+					err = json.NewDecoder(ctx.Req.Request.Body).Decode(v)
+				}
+			}
+			if err == nil && ctx.Req.URL != nil {
```

Comment:
Why this `if` part code block is necessary?

Reply:
For POST/PUT/PATCH verb, I am preserving the original error. Also, `ctx.Req.URL` is a pointer, so just checking !=nil for safety.

I am somewhat following the patter here: https://github.com/golang/go/blob/master/src/net/http/request.go#L1194

Reply:
Hmm, I mean the body of this whole block, what it is doing? (not the if check condition)

Reply:
I have updated the pr.